src/helper/sampler.hpp

```diff
@@ -33,7 +33,7 @@ class sampler {
     public:
         explicit scope(std::atomic<bool> &concurrent) : concurrent_(&concurrent)
         {
-            *concurrent_ = true;
+            concurrent_->store(true, std::memory_order_relaxed);
```

Comment:
Why do you have this `concurrent_` variable in the first place? it seems that its purpose is to prevent two `scope` objects to exist simultaneously (in a bit a confusing way, because it's not immediately clear from the code and it relies on another mutex to ensure this property, rather than just using  plain CAS). But the result is that the true sampling rate will be smaller than the requested one; isn't this a spec violation?

Reply:
This roughly prevents two requests from engaging the schema extraction processor at the same time, this was specified in the RFC as a way to reduce the performance impact of the feature, so not a spec violation. 

On the other hand, the sampling rate specified is mostly for guidance and we expect the libraries to make any required compromises to avoid impacting endpoints with heavy load.

Reply:
Ok I was not aware of this, I don't think this was a thing when I looked in Java. But anyway regardless of the merits, the way it's implemented is difficult to reason about because of the interaction between the mutex and the atomic variable. This could be done in a simpler and more efficient way without the mutex. Use fetch_and_add (weak) for the request count and compare_exchange_weak on the concurrent flag.